internal/hcp/internal/controllers/link/controller.go

```diff
@@ -43,33 +42,28 @@ func LinkController(
 	hcpAllowV2ResourceApis bool,
 	hcpClientFn HCPClientFn,
 	cfg config.CloudConfig,
-	dataDir string,
 	hcpManager hcp.Manager,
 ) *controller.Controller {
 	return controller.NewController("link", pbhcp.LinkType).
-		// Placement is configured to each server so that the HCP manager is started
-		// on each server. We plan to implement an alternative strategy to starting
-		// the HCP manager so that the controller placement will eventually only be
-		// on the leader.
-		// https://hashicorp.atlassian.net/browse/CC-7364
-		WithPlacement(controller.PlacementEachServer).
-		WithInitializer(&linkInitializer{
-			cloudConfig: cfg,
-		}).
-		WithReconciler(&linkReconciler{
-			resourceApisEnabled:    resourceApisEnabled,
-			hcpAllowV2ResourceApis: hcpAllowV2ResourceApis,
-			hcpClientFn:            hcpClientFn,
-			dataDir:                dataDir,
-			hcpManager:             hcpManager,
-		})
+		WithInitializer(
+			&linkInitializer{
+				cloudConfig: cfg,
+			},
+		).
+		WithReconciler(
+			&linkReconciler{
+				resourceApisEnabled:    resourceApisEnabled,
+				hcpAllowV2ResourceApis: hcpAllowV2ResourceApis,
+				hcpClientFn:            hcpClientFn,
+				hcpManager:             hcpManager,
```

Comment:
I was thinking it would be nice if we didn’t have to pass the HCP manager to the controller and keep those two concepts separate.

Since we’re now only using it to get the current cloud config, maybe we should instead merge the incoming link values with the initial cloud config that we currently use in the initializer. The behavior would technically be different but it would accomplish the same end goal (preserve node information and any HCP connection details that might have been configured at startup).

If we went with this change, we would want to pass the initial cloud config to the [`MonitorHCPLink`](https://github.com/hashicorp/consul/pull/20401/files#diff-3ca707dfa544d1d3dcb81a59461f74d125823328fe8545c311b16a0fdb6ef9b3R63-R74) as well so that we merge the client config the same way.

Reply:
Totally agree. I was thinking the same thing, but wasn't sure if this PR was getting too big. If you think it makes sense to do here, I'll make that change.

Reply:
This is done